authentication/azure_sp_objectid.go

```diff
@@ -3,47 +3,144 @@ package authentication
 import (
 	"context"
 	"fmt"
+	"log"
+	"net/http"
+	"net/http/httputil"
+	"strings"
 
 	"github.com/Azure/azure-sdk-for-go/services/graphrbac/1.6/graphrbac"
+	"github.com/manicminer/hamilton/auth"
+	"github.com/manicminer/hamilton/environments"
+	"github.com/manicminer/hamilton/msgraph"
+	"github.com/manicminer/hamilton/odata"
+
 	"github.com/hashicorp/go-azure-helpers/sender"
 )
 
 func buildServicePrincipalObjectIDFunc(c *Config) func(ctx context.Context) (string, error) {
 	return func(ctx context.Context) (string, error) {
-		env, err := AzureEnvironmentByNameFromEndpoint(ctx, c.MetadataHost, c.Environment)
-		if err != nil {
+		objectId, err := objectIdFromMsGraph(ctx, c)
+		if err == nil {
+			return objectId, nil
+		}
+		if !strings.HasPrefix(err.Error(), "access denied") {
 			return "", err
 		}
 
-		s := sender.BuildSender("GoAzureHelpers")
+		log.Printf("Unable to retrieve service principal object ID from Microsoft Graph: %+v", err)
+		return objectIdFromAadGraph(ctx, c)
+	}
+}
 
-		oauthConfig, err := c.BuildOAuthConfig(env.ActiveDirectoryEndpoint)
-		if err != nil {
-			return "", err
-		}
+func objectIdFromAadGraph(ctx context.Context, c *Config) (string, error) {
+	env, err := AzureEnvironmentByNameFromEndpoint(ctx, c.MetadataHost, c.Environment)
+	if err != nil {
+		return "", err
+	}
 
-		// Graph Endpoints
-		graphEndpoint := env.GraphEndpoint
-		graphAuth, err := c.GetAuthorizationToken(s, oauthConfig, env.GraphEndpoint)
-		if err != nil {
-			return "", err
-		}
+	s := sender.BuildSender("GoAzureHelpers")
+
+	oauthConfig, err := c.BuildOAuthConfig(env.ActiveDirectoryEndpoint)
+	if err != nil {
+		return "", err
+	}
 
-		client := graphrbac.NewServicePrincipalsClientWithBaseURI(graphEndpoint, c.TenantID)
-		client.Authorizer = graphAuth
-		client.Sender = s
+	graphAuth, err := c.GetAuthorizationToken(s, oauthConfig, env.GraphEndpoint)
+	if err != nil {
+		return "", err
+	}
 
-		filter := fmt.Sprintf("appId eq '%s'", c.ClientID)
-		listResult, listErr := client.List(ctx, filter)
+	client := graphrbac.NewServicePrincipalsClientWithBaseURI(env.GraphEndpoint, c.TenantID)
+	client.Authorizer = graphAuth
+	client.Sender = s
 
-		if listErr != nil {
-			return "", fmt.Errorf("Error listing Service Principals: %#v", listErr)
-		}
+	filter := fmt.Sprintf("appId eq '%s'", c.ClientID)
+	listResult, listErr := client.List(ctx, filter)
+
+	if listErr != nil {
+		return "", fmt.Errorf("listing Service Principals: %#v", listErr)
+	}
+
+	if listResult.Values() == nil || len(listResult.Values()) != 1 || listResult.Values()[0].ObjectID == nil {
+		return "", fmt.Errorf("unexpected Service Principal query result: %#v", listResult.Values())
+	}
+
+	return *listResult.Values()[0].ObjectID, nil
+}
+
+func objectIdFromMsGraph(ctx context.Context, c *Config) (string, error) {
```

Comment:
can we return a pointer here to ensure we don't use an invalid value accidentally (means we'll crash if we attempt to run without handling the error, which'll get caught at acctest time):

```suggestion
func objectIdFromMsGraph(ctx context.Context, c *Config) (*string, error) {
```